text/0021-any-and-all-operators.md

```diff
@@ -0,0 +1,165 @@
+
+2. Predicates cannot contain nested `any` or `all` expressions. We disallow nested quantifiers for both performance and analyzability reasons. Nested quantifiers allow writing expressions that run in $O(2^n)$ time, where $n$ is the number of nested quantifiers. For example:
+
+```
+[0,1].all(
+  [0,1].all(
+    ...
+      [0,1].all(true)...))
+```
+
+Because nested quantifiers are prohibited, there is no need to support predicates with arbitrary parameter names. The `it` keyword sufficiently extends the grammar.
+
+### Extending the semantics
+
+The semantics of `all` and `any` are straightforward when there are no errors:
+- `E.all(P)` evaluates to true when `P[e/it]` is true for every element `e` in `E`.
+- `E.any(P)` evaluates to true when `P[e/it]` is true for some element `e` in `E`.
+
+But what happens if evaluating `P` errors on some element `e`? In this case, the evaluation of `all` or `any` terminates abruptly with a distinguished `QuantifierError`.
+
+The `QuantifierError` error may include static information (such as source location), but no dynamic information (such as the value that caused the error to be thrown). This error handling approach ensures that evaluating  `all` and `any` is _deterministic_: the result is the same regardless of iteration order.
+
+For example, consider the expression `[1, true].all(it like "foo")`. Regardless of the order in which the predicate is applied, the expression returns the same `QuantifierError`.
+
+This semantics lets us think about `all` and `any` in terms of their expansion to conjunctions and disjunctions. The quantified expression errors if and only if the expansion would error.
+
+For example:
+```
+// Throws QuantifierError
+[1, true].all(it like "foo")
+
+// Throws a TypeError: "like" can't be applied to an integer
+1 like "foo" && true like "foo"
+
+// Throws a TypeError: "like" can't be applied to an boolean
+true like "foo" && 1 like "foo"
+```
+
+The only difference is the error being thrown: `QuantifierError` for the quantified expression versus an element-specific error for the explicit expansion.
+
+### Extending the type system
+
+The type system needs to be extended with rules for type checking quantified expressions `E.all(P)` and `E.any(P)`. This should be fairly standard: we know the type of `it` from the type of the set `E`, and we can use this to typecheck the predicate `P`. We also need to propagate effects through predicates.
+
+For example, the following should typecheck when `context.limit` is an optional integer attribute, and `resource.things` is a set of entities with an optional integer attribute `.size`:
+
+```
+context has limit &&
+resource.things.all(
+  it has size && it.size < context.limit
+)
```

Comment:
While the effect prior to a `.any` or `.all` is clear, what about the effect after one? The simplest thing would be to ignore the effects, but we could be more precise. For example
```
resource.things.all(it has size && it.size < 5) &&
context has limit &&
resource.things.all(it.size < context.limit)
```
The second use of `all` could benefit from the first, which has confirmed that all the elements of `resource.things` have the `size` attribute. A simpler example would be
```
resource.things.all(context has limit && it.size < context.limit) &&
context.limit > 5
```
Here, the redundant use of `context.limit` in the `all` carries an effect outside to the `>` expression.

Reply:
Good callout.  Yes, this would increase precision, though it might not be worth the extra complication in the implementation.  For these examples at least, there are natural ways to rewrite them so that don't rely on the extra precision.  For example, the first expression could be rewritten as

```
context has limit &&
resource.things.all(it has size && it.size < 5 && it.size < context.limit)
```

Reply:
I don't see a natural way to rewrite when the second quantifier is an `any`
```
context has limit &&
resource.things.all(it has size && it.size < 5) &&
resource.things.any(it.size < context.limit)
```